### PR TITLE
lsp: use `detail` in qf-list for call hierarchies

### PR DESCRIPTION
**Pull request by cbarrete**
`detail` will contain more information about the function, typically its
signature, which is more useful than only its name.

### CODE REVIEW

File: runtime/lua/vim/lsp/handlers.lua [hunk 1]
```diff
@@ -379,7 +379,7 @@ local make_call_hierarchy_handler = function(direction)
       for _, range in pairs(call_hierarchy_call.fromRanges) do
         table.insert(items, {
           filename = assert(vim.uri_to_fname(call_hierarchy_item.uri)),
-          text = call_hierarchy_item.name,
+          text = call_hierarchy_item.detail,
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/handlers.lua:382]:
```suggestion
          text = call_hierarchy_item.detail or call_hierarchy_item.name,
```

According to the specification, `detail` is optional

Comment (cbarrete) [reply runtime/lua/vim/lsp/handlers.lua]: Good catch, thanks!

**Comment by cbarrete** [comment]
@tjdevries Would you mind having a look? This shouldn't be too controversial.

**Comment by cbarrete** [comment]
@mjlbach instead maybe? :)

**Comment by mjlbach** [comment]
@cbarrete what's a language server/file that I can test this with? I never use call hierarchies

**Comment by cbarrete** [comment]
With `rust-analyzer`:

```rustfn sum(x: i8, y: i8) -> i8 {
    x + y
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    sum(1, 2);
    Ok(())
}
```

You can `:lua vim.lsp.buf.outgoing_calls()` over `main` and `:lua vim.lsp.buf.incoming_calls()` over `sum`.
Notice the function signatures instead of just their names in the qflist.

**Comment by mjlbach** [comment]
Can you rebase to fix test failure?

**Comment by mjlbach** [comment]
Also do you know the behavior of vscode? I tried to use RA on vscode but I get an error (Command 'Calls: Show Incoming Calls' resulted in an error (command 'references-view.showIncomingCalls' not found)

**Comment by cbarrete** [comment]
> Also do you know the behavior of vscode? I tried to use RA on vscode but I get an error (Command 'Calls: Show Incoming Calls' resulted in an error (command 'references-view.showIncomingCalls' not found)

No I don't, although I'd expect it to work since VSC is the best supported editor by RA

**Comment by mfussenegger** [comment]
With eclipse.jdt.ls`  `detail` contains the full class name and `name` contains the method name + type:

```
"item", {
  detail = "org.elasticsearch.cluster.service.ClusterService",
  kind = 6,
  name = "getClusterSettings() : ClusterSettings",
```

Should we include both, name and detail?

**Comment by cbarrete** [comment]
I've digged around a little and here's what I've found:
- The spec is vague enough ("More detail for this item, e.g. the signature of a function.") that neither servers or clients agree on how to use `detail`
- Not all servers that support call hierarchies fill in the `detail` field
- Those that do put completely different things in there (e.g. jdt ls vs rust analyzer)
- VSCode displays both `name` and `detail`, one after the other

In the RA case, it looks odd to have e.g. `f | f(x: i8)` in the qflist, but that's something that can be modified on their side (it would look better in VSCode as well so it should be easy to convince enough them).

I think that it's not worth having a configuration option here, we should print the name and optionally the detail if it's there as well.

One thing I'd be happy to hear about though is what everyone thinks it should look like. The only relevant qflist field here is `text`, so we'll need some way of separating name/detail (I don't think it's worth extending the qflist just for that).
Any suggestions?

**Comment by mjlbach** [comment]
I believe we should match the vscode behavior, as it is most likely that if servers are going to change what is returned for name/detail, they will do  so in response primarily due to users using the vscode extensions. I wasn't able to get the call hierarchy working for vscode (I didn't try especially hard though), is this universal across all LSP extensions?

**Comment by cbarrete** [comment]
>I wasn't able to get the call hierarchy working for vscode (I didn't try especially hard though)

Weird, I just installed VSCode + the rust-analyzer extension and it just worked.

>is display name + detail universal across all LSP extensions?

Not sure what you mean by this. The way they are displayed is dictated by the editor, so they look the same, but the contents might be different (again, because they are loosely specified).

**Pull request closed without merging**